rust/connlib/tunnel/src/client.rs

```diff
@@ -117,8 +122,18 @@ where
     #[tracing::instrument(level = "trace", skip(self))]
     pub fn set_interface(&mut self, config: InterfaceConfig) -> connlib_shared::Result<()> {
         self.role_state.interface_config = Some(config);
-        self.update_dns_mapping();
-        self.update_interface()
+        let dns_changed = self.role_state.update_dns_mapping();
+
+        if dns_changed {
+            self.io
+                .set_upstream_dns_servers(self.role_state.dns_mapping());
+
+            tracing::info!("Setting new DNS resolvers");
+        }
```

Comment:
This is a bit odd but in fact the behaviour we currently have. At the moment, we ignore the return value from `update_dns_mapping` in `set_interface`. But `update_dns_mapping` has a shortcut internally to only update the upstream DNS servers when the DNS servers actually changed.

However, regardless of whether they changed, we currently call `update_interface`. I guess that is because the interface config itself changed.

I think it makes sense but it wasn't at all obvious to me with how the code was previously written.